src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -6266,13 +6335,135 @@ private TypeSyntax ParseUnderlyingType(bool parentIsParameter, NameOptions optio
+                var greaterThanTokenError = WithAdditionalDiagnostics(SyntaxFactory.MissingToken(SyntaxKind.GreaterThanToken), GetExpectedTokenError(SyntaxKind.GreaterThanToken, SyntaxKind.None));
+
+                var funcPtr = SyntaxFactory.FunctionPointerType(@delegate, asterisk, callingConvention, lessThanTokenError, typesError, greaterThanTokenError);
+                _pool.Free(typesError);
+                return funcPtr;
+            }
+
+            var lessThanToken = EatToken(SyntaxKind.LessThanToken);
+            var saveTerm = _termState;
+            _termState |= TerminatorState.IsEndOfFunctionPointerParameterList;
+            var types = _pool.AllocateSeparated<ModifiedType>();
+
+            try
+            {
+                while (true)
+                {
+                    var modifiers = _pool.Allocate<SyntaxToken>();
+                    try
+                    {
+                        while (IsFunctionPointerParameterModifier())
+                        {
+                            modifiers.Add(EatToken());
+                        }
+
+                        var parameterType = ParseTypeOrVoid();
+                        types.Add(SyntaxFactory.ModifiedType(modifiers, parameterType));
+
+                        if (skipBadFunctionPointerParameterListTokens() == PostSkipAction.Abort)
+                        {
+                            break;
+                        }
+
+                        Debug.Assert(CurrentToken.Kind == SyntaxKind.CommaToken);
+                        types.AddSeparator(EatToken(SyntaxKind.CommaToken));
+                    }
+                    finally
+                    {
+                        _pool.Free(modifiers);
+                    }
+                }
+
+                var greaterThanToken = EatToken(SyntaxKind.GreaterThanToken);
+                var funcPointer = SyntaxFactory.FunctionPointerType(@delegate, asterisk, callingConvention, lessThanToken, types, greaterThanToken);
+                funcPointer = CheckFeatureAvailability(funcPointer, MessageID.IDS_FeatureFunctionPointers);
+                return funcPointer;
+            }
+            finally
+            {
+                _termState = saveTerm;
+                _pool.Free(types);
+            }
+
+            PostSkipAction skipBadFunctionPointerParameterListTokens()
+            {
+                CSharpSyntaxNode? tmp = null;
+                Debug.Assert(types.Count > 0);
+                return SkipBadSeparatedListTokensWithExpectedKind(ref tmp,
+                    types,
+                    isNotExpectedFunction: p => p.CurrentToken.Kind != SyntaxKind.CommaToken,
+                    abortFunction: p => p.IsTerminator(),
+                    expected: SyntaxKind.CommaToken);
+            }
+        }
+
+        private bool IsFunctionPointerParameterModifier()
+            => IsParameterModifier(CurrentToken.Kind) || CurrentToken.Kind == SyntaxKind.ReadOnlyKeyword;
```

Comment:
>CurrentToken.Kind == SyntaxKind.ReadOnlyKeyword [](start = 57, length = 47)

Is there some magic around this token?

Reply:
`readonly` isn't a valid regular parameter modifier, but it's allowed on `ref readonly` returns, so it needs to be supported here.

---
In reply to: [349361489](https://github.com/dotnet/roslyn/pull/39879#discussion_r349361489) [](ancestors = 349361489)

Reply:
> readonly isn't a valid regular parameter modifier, but it's allowed on ref readonly returns, so it needs to be supported here.

It doesn't have to be preceded by ```ref```? Could you give an example?

---
In reply to: [349363828](https://github.com/dotnet/roslyn/pull/39879#discussion_r349363828) [](ancestors = 349363828,349361489)

Reply:
It does indeed have to be preceded by `ref`, but I was planning validating the order and presence of modifiers in the type binding step.

---
In reply to: [349370516](https://github.com/dotnet/roslyn/pull/39879#discussion_r349370516) [](ancestors = 349370516,349363828,349361489)

Reply:
> It does indeed have to be preceded by ref, but I was planning validating the order and presence of modifiers in the type binding step.

Why differ from the way we deal with modifiers in regular signatures. And if this is not different, can we share the logic?

---
In reply to: [349371508](https://github.com/dotnet/roslyn/pull/39879#discussion_r349371508) [](ancestors = 349371508,349370516,349363828,349361489)

Reply:
> Why differ from the way we deal with modifiers in regular signatures

This isn't particularly different. The only validation that `ParseParameterModifiers` does at this stage, for example, is language version checks. I could add a flag to that function that allows it to check for `readonly` as well, as it doesn't handle them currently, but I decided that there wasn't a huge benefit to doing this. I can modify it if you think it's really warranted though.

---
In reply to: [349378051](https://github.com/dotnet/roslyn/pull/39879#discussion_r349378051) [](ancestors = 349378051,349371508,349370516,349363828,349361489)

Reply:
Followed up offline. This isn't quite the same as a regular parameter modifier because it also allows `ref readonly`, where regular parameters do not.

---
In reply to: [349741647](https://github.com/dotnet/roslyn/pull/39879#discussion_r349741647) [](ancestors = 349741647,349378051,349371508,349370516,349363828,349361489)

Reply:
That being said, if we ever want to add new modifiers to regular parameters we'll need to do language version checks, so if we have `ParseParameterModifiers` take a flag and understand this, it would probably good for future proofing.

---
In reply to: [349809401](https://github.com/dotnet/roslyn/pull/39879#discussion_r349809401) [](ancestors = 349809401,349741647,349378051,349371508,349370516,349363828,349361489)